configstack/module.go

```diff
@@ -52,7 +53,68 @@ func ResolveTerraformModules(terragruntConfigPaths []string, terragruntOptions *
 		return []*TerraformModule{}, err
 	}
 
-	return crosslinkDependencies(mergeMaps(modules, externalDependencies), canonicalTerragruntConfigPaths)
+	crossLinkedModules, err := crosslinkDependencies(mergeMaps(modules, externalDependencies), canonicalTerragruntConfigPaths)
+	if err != nil {
+		return []*TerraformModule{}, err
+	}
+
+	finalModules := removeExcludedDirs(crossLinkedModules, terragruntOptions)
+
+	return finalModules, nil
+}
+
+// removeExcludedDirs iterates over a module slice and removes all entries, which should be ignored via the terragrunt-exclude-dir CLI flag.
+// Modules are also removed if one of its dependencies are to be excluded. Returns the cleaned-up slice of modules.
+func removeExcludedDirs(modules []*TerraformModule, terragruntOptions *options.TerragruntOptions) []*TerraformModule {
+
+	// If no excludeDir is specified return the modules list instantly
+	if terragruntOptions.ExcludeDir == "" {
+		return modules
+	}
+
+	excludeGlobMatches := []string{}
+
+	// If possible, expand the glob to get all excluded filepaths
+	excludeGlobMatches, err := zglob.Glob(terragruntOptions.ExcludeDir)
+
+	if err != nil {
+		// Glob can not be expanded, using literal value
+		excludeGlobMatches = []string{terragruntOptions.ExcludeDir}
+	}
+
+	// If the expanded glob paths are relative, generate absolute paths
+	absoluteExludeDir := []string{}
+	for _, m := range excludeGlobMatches {
+		if strings.HasPrefix(m, "/") {
+			absoluteExludeDir = append(absoluteExludeDir, m)
+		} else {
+			absoluteExludeDir = append(absoluteExludeDir, fmt.Sprintf("%s/%s", terragruntOptions.WorkingDir, m))
+		}
+	}
+
+	// Create a slice of finalModules to store all modules, which should not be excluded
+	finalModules := modules[:0]
+
+	for _, e := range absoluteExludeDir {
+		for _, m := range modules {
+
+			// Check if none of the current modules dependencies are located under the excluded directory
```

Comment:
Hm, so should a module `A` be excluded just because it depends on a module `B` that is excluded? I understand that if `B` has never been applied, then `A` might fail, but what if `B` is something like a VPC that has been applied, long ago, and you just don't want to (uselessly) keep applying it again and again?

Reply:
This is a very good point. I think it is safe to assume, that if the user excplicitly excludes a directory, it can be assumed that all other modules will still be applied, even if there dependencies are not fulfilled. I can adjust the code to reflect this behavior, what do you think?

Reply:
Yea, I think we should only exclude the directories the user explicitly tells us to and not their dependencies.

Reply:
@brikis98 I adapted this so we are no longer excluding the modules which depend on excluded modules. However I am now running in a `DependencyNotFoundWhileCrossLinking` error. 

I tracked it down to `crossLinkDependencies` inside `running_module.go`.

This error appears because it tries to cross-link dependencies during the run, and the modules dependencies are no longer in the list of modules.

Is the best solution here, to remove the affected dependencies in `removeExcludedDirs`?

Reply:
FYI, I did not push that code yet, as it contains a lot of debug output for tests.

Reply:
Hm, is it possible to remove excluded modules after the call to `crossLinkDependencies` in `running_module.go`?

Reply:
you mean instead of removing them in `module.go`? 
I assume this would be in `toRunningModules()` then? However I am missing the terragrunt options there and I am not yet completely sure if this is the correct place to do it.

Also, why are dependencies linked here again, when this is already done in `ResolveTerraformModules`?

I would need some guidance here on where best to remove the excluded modules. Would be great if you could give some more insights.

Reply:
I fixed this now by introducing another field called `flagExcluded` to the module struct.
This flag is set in `flagExcludedDirs` in `module.go` as all necessary information is available here.

The modules and the affected dependencies are then later actually removed in `crossLinkDependencies` in `running_module.go`.

This was the cleanest way of making this work, as I was missing essential information inside `running_module.go`.